### PR TITLE
Issue #SB-867 feat: Refactoring announcement feature

### PR DESCRIPTION
**Pull request by niteshkesarkar**

### CODE REVIEW

File: src/app/private/scripts/controllers/announcement/createAnnouncementController.js [hunk 1]
```diff
@@ -177,133 +143,110 @@ angular.module('playerApp').controller('createAnnouncementCtrl', ['$rootScope',
...
-          createAnn.showError(apiResponse)
-        }
-      }).catch(function (apiResponse) {
-        createAnn.isMetaModified = true
-        createAnn.showError(apiResponse.data)
-      })
+          $state.go('announcementOutbox')
+        }, function (err) {
+          composeAnn.isMetaModified = true
+          composeAnn.showError(apiResponse.data)
```
[the remark this answers is not in the document]
Comment (niteshkesarkar) [reply src/app/private/scripts/controllers/announcement/createAnnouncementController.js]: resolved

File: src/test/spec/app/private/scripts/controllers/announcement/resendAnnouncementController.js [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+     */
+    composeAnn.removeRecipients = function (item) {
+      _.remove(composeAnn.announcement.target.geo.ids, function (arg) {
+        if (arg == item.id) {
+          return true
+        }
+      })
+      composeAnn.announcement.target.geo.ids
+      _.remove(composeAnn.announcement.selTar, function (arg) {
+        if (arg.location == item.location) {
```
[the remark this answers is not in the document]
Comment (niteshkesarkar) [reply src/test/spec/app/private/scripts/controllers/announcement/resendAnnouncementController.js]: resolved

File: src/app/common/js/frmelmnts.js [hunk 3]
```diff
@@ -1,3 +1,3 @@
...
...r":"Phone Number","dob":"Date Of Birth","lang":"Language","languages":"Bengali,English,Gujarati,Hindi,Kannada,Marathi,Punjabi,Tamil,Telugu","page":"Page","tryagain":"Try Again","close":"Close","createnewanncmnt":"Create New Announcement","createnewanncmntdesc":"Fill in all the required fields to create the announcement.","resendanncmntdesc":"You can make any edits required and resend the announcement.","selectrecipient":"Select Announcement Recipients","selectrecipientdesc":"Select groups who should receive the announcement.","editrecipientdesc":"Edit groups who should receive the announcement.","confirmrecipient":"Confirm Announcement Recipients","confirmrecipientdesc":"Review and confirm your announcement recipients.","previewanncmnt":"Preview Announcement","previewanncmntdesc":"This is how your announcement will look.","anncmntstep":"STEP","anncmnttitle":"TITLE","anncmntfrom":"FROM","anncmntfromDesc":"Please enter organisation/department that has sent the announcement","anncmnttype":"ANNOUNCEMENT TYPE","anncmntdesc":"DESCRIPTION (OPTIONAL)","anncmntlink":"URL / WEB LINK","anncmntaddlink":"Add URL/web link","anncmntaddlinkdesc":"(For example, you can link a Google Form)","anncmnt
```
Comment (kochhar) [inline src/app/common/js/frmelmnts.js]: please reformat this line -- it's too long.

File: src/app/private/scripts/controllers/announcement/announcementInboxListController.js [hunk 4]
```diff
@@ -10,21 +10,20 @@ angular.module('playerApp')
        * @method renderAnnouncementList
        * @desc - function to render inbox announcement
        * @memberOf Controllers.announcementInboxListController
-       * @param {string} [user id]
        * @param {int} [limit]
        */
       announcementInboxData.renderAnnouncementList = function (limit) {
         announcementInboxData.limit = limit || -1
-        announcementAdapter.getInboxAnnouncementList($rootScope.userId, announcementInboxData.limit).then(function (apiResponse) {
+        announcementAdapter.getInboxAnnouncementList(announcementInboxData.limit).then(function (apiResponse) {
```
Comment (kochhar) [inline src/app/private/scripts/controllers/announcement/announcementInboxListController.js]: please reformat.

File: src/app/private/scripts/controllers/announcement/announcementInboxListController.js [hunk 5]
```diff
@@ -10,21 +10,20 @@ angular.module('playerApp')
...
           announcementInboxData.listData = apiResponse.result.announcements
 
           // Converting attachment to object from string
           _.forEach(announcementInboxData.listData, function (announcement) {
             // Call received API
             if (announcement.received === false) {
-              announcementAdapter.receivedAnnouncement($rootScope.userId, announcement.id).then(function (response) {
-                console.log('Received success')
+              announcementAdapter.receivedAnnouncement(announcement.id).then(function (response) {
+                //console.log('Received success')
```
Comment (kochhar) [inline src/app/private/scripts/controllers/announcement/announcementInboxListController.js]: please delete.

File: src/app/private/scripts/controllers/announcement/announcementOutboxListController.js [hunk 6]
```diff
@@ -15,10 +15,10 @@ angular.module('playerApp')
      * @memberOf Controllers.announcementOutboxListController
      */
       announcementOutboxData.renderAnnouncementList = function () {
-        announcementAdapter.getOutBoxAnnouncementList($rootScope.userId).then(function (apiResponse) {
+        announcementAdapter.getOutBoxAnnouncementList().then(function (apiResponse) {
```
Comment (kochhar) [inline src/app/private/scripts/controllers/announcement/announcementOutboxListController.js]: please reformat.

File: src/app/private/scripts/controllers/announcement/announcementOutboxListController.js [hunk 7]
```diff
@@ -82,8 +82,9 @@ angular.module('playerApp')
      * @memberOf Controllers.announcementOutboxListController
      */
       announcementOutboxData.deleteAnnouncement = function () {
-        announcementAdapter.deleteAnnouncement($rootScope.userId, announcementOutboxData.announcementId).then(function (apiResponse) {
+        announcementAdapter.deleteAnnouncement(announcementOutboxData.announcementId).then(function (apiResponse) {
```
Comment (kochhar) [inline src/app/private/scripts/controllers/announcement/announcementOutboxListController.js]: please reformat.

File: src/app/private/scripts/controllers/announcement/announcementOutboxListController.js [hunk 8]
```diff
@@ -104,12 +105,15 @@ angular.module('playerApp')
...
-      announcementOutboxData.getResend = function (announcementId) {
-        announcementAdapter.getResend(announcementId).then(function (apiResponse) {
-          if (apiResponse.hasOwnProperty('result')) {
-            $rootScope.$broadcast('editAnnouncementBeforeResend', apiResponse.result)
-          }
-        })
+      announcementOutboxData.getResend = function (announcementId, announcementTitle) {
+        var hashTagId = ''
+        if ($stateParams.userIdHashTag === undefined) {
+          hashTagId = (Math.floor(new Date().getTime() / 1000)) + ($rootScope.userId) + (Math.floor(Math.random() * 90000) + 10000)
```
Comment (kochhar) [inline src/app/private/scripts/controllers/announcement/announcementOutboxListController.js]: please fix line length.
please fix random number generation.

File: src/app/private/scripts/controllers/announcement/announcementOutboxListController.js [hunk 9]
```diff
@@ -104,12 +105,15 @@ angular.module('playerApp')
...
+          hashTagId = md5(hashTagId)
+        } else {
+          hashTagId = $stateParams.userIdHashTag
+        }
+        $state.go('announcementResend', {announcementId: announcementId, stepNumber: '1', userIdHashTag: hashTagId, telemetryAnnTitle: announcementTitle})
```
Comment (kochhar) [inline src/app/private/scripts/controllers/announcement/announcementOutboxListController.js]: please fix line length.

File: src/app/private/scripts/controllers/announcement/announcementOutboxListController.js [hunk 10]
```diff
@@ -121,5 +125,21 @@ angular.module('playerApp')
...
         $state.go('announcementDetails', {announcementId: annId, announcementName: item.details.title, pageId: 'announcement_outbox_view'})
       }
+
+    /**
+     * @method gotToAnnouncementCreateState
+     * @desc - function to change the current state to announcement create
+     * @memberOf Controllers.announcementOutboxListController
+     */
+      announcementOutboxData.gotToAnnouncementCreateState = function () {
+        var hashTagId = ''
```
Comment (kochhar) [inline src/app/private/scripts/controllers/announcement/announcementOutboxListController.js]: code is duplicated. please remove duplication.

File: src/app/private/scripts/controllers/announcement/createAnnouncementController.js [hunk 11]
```diff
@@ -1,65 +1,43 @@
...
-    var getDefinitionReq = {
-      request: {
-        'rootorgid': $rootScope.rootOrgId,
-        'userid': $rootScope.userId,
-        'definitions': ['announcementtypes', 'senderlist']
+angular.module('playerApp').controller('createAnnouncementCtrl', ['$rootScope', '$scope', '$state', '$stateParams', '$timeout', 'config', 'toasterService', 'announcementService', 'fileUpload', 'AnnouncementModel', 'announcementAdapter', 'portalTelemetryService',
+  function ($rootScope, $scope, $state, $stateParams, $timeout, config, toasterService, announcementService, fileUpload, AnnouncementModel, announcementAdapter, portalTelemetryService) {
+    var composeAnn = this
+    composeAnn.data = {}
+    composeAnn.attachment = []
```
Comment (kochhar) [inline src/app/private/scripts/controllers/announcement/createAnnouncementController.js]: these fields should not be required. they are provided in the announcement model.

File: src/app/private/scripts/controllers/announcement/createAnnouncementController.js [hunk 12]
```diff
@@ -68,19 +46,28 @@ angular.module('playerApp').controller('createAnnouncementCtrl', ['$rootScope',
...
+            composeAnn.isMetaModified = true
+          }
+
+          if (composeAnn.isMetaModified === true && composeAnn.isMetaModifiedSteps !== true) {
+            composeAnn.confirmationModal()
+            return false
+          } else if (composeAnn.isMetaModified === false && composeAnn.stepNumber === 1) {
+            composeAnn.refreshFormValues()
+            $state.go('announcementOutbox')
+            // composeAnn.hideModel('createAnnouncementModal')
```
Comment (kochhar) [inline src/app/private/scripts/controllers/announcement/createAnnouncementController.js]: please delete commented code.

File: src/app/private/scripts/controllers/announcement/createAnnouncementController.js [hunk 13]
```diff
@@ -91,16 +78,18 @@ angular.module('playerApp').controller('createAnnouncementCtrl', ['$rootScope',
...
           allowMultiple: true,
           onDeny: function () {
             return true
           },
           onApprove: function () {
-            createAnn.refreshFormValues()
-            createAnn.hideModel('announcementCancelModal')
+            composeAnn.isApprove = true
+            composeAnn.refreshFormValues()
+            composeAnn.hideModel('announcementCancelModal')
```
Comment (kochhar) [inline src/app/private/scripts/controllers/announcement/createAnnouncementController.js]: function name should be `hideModal` not `hideModel`

File: src/app/private/scripts/controllers/announcement/createAnnouncementController.js [hunk 14]
```diff
@@ -125,12 +114,12 @@ angular.module('playerApp').controller('createAnnouncementCtrl', ['$rootScope',
      * @desc - add new url input box
      * @memberOf Controllers.createAnnouncementCtrl
      */
-    createAnn.addNewLink = function () {
-      var newItemNo = createAnn.repeatableWebLinks.length + 1
-      createAnn.repeatableWebLinks.push({
+    composeAnn.addNewLink = function () {
+      var newItemNo = composeAnn.repeatableWebLinks.length + 1
+      composeAnn.repeatableWebLinks.push({
```
Comment (kochhar) [inline src/app/private/scripts/controllers/announcement/createAnnouncementController.js]: this state change should be made on the AnnouncementModel not on the controller state. please fix.

File: src/app/private/scripts/controllers/announcement/createAnnouncementController.js [hunk 15]
```diff
@@ -139,36 +128,13 @@ angular.module('playerApp').controller('createAnnouncementCtrl', ['$rootScope',
...
-        })
-      }
-      createAnn.previewData = {
-        'details': { 'type': createAnn.data.type, 'title': createAnn.data.title, 'description': createAnn.data.description, 'from': createAnn.data.from },
-        'sourceid': $rootScope.rootOrgId,
-        'links': createAnn.linkArray,
-        'target': ['teachers'],
-        'attachments': createAnn.attachment
+    composeAnn.removeLink = function (index) {
+      composeAnn.repeatableWebLinks.splice(index, 1)
```
Comment (kochhar) [inline src/app/private/scripts/controllers/announcement/createAnnouncementController.js]: this state change should be made on the AnnouncementModel class not on the controller state.

File: src/app/private/scripts/controllers/announcement/createAnnouncementController.js [hunk 16]
```diff
@@ -177,133 +143,110 @@ angular.module('playerApp').controller('createAnnouncementCtrl', ['$rootScope',
...
-        if (apiResponse && apiResponse.responseCode === 'OK') {
-          $timeout(function () {
-            createAnn.refreshFormValues()
-          })
+    composeAnn.saveAnnouncement = function () {
+      composeAnn.isMetaModifiedSteps = true
+      composeAnn.announcement.target.geo.ids = _.map(composeAnn.announcement.selTar, 'id')
+      announcementAdapter.createAnnouncement(composeAnn.announcement)
+        .then(function (apiResponse) {
+          composeAnn.hideModel('createAnnouncementModal')
```
Comment (kochhar) [inline src/app/private/scripts/controllers/announcement/createAnnouncementController.js]: function name should be `hideModal` not `hideModel`

File: src/app/private/scripts/controllers/announcement/createAnnouncementController.js [hunk 17]
```diff
@@ -345,17 +288,24 @@ angular.module('playerApp').controller('createAnnouncementCtrl', ['$rootScope',
...
         fileSizeErrorText: $rootScope.messages.imsg.m0021,
         containerName: 'attachments/announcement',
-        uploadSuccess: createAnn.onUploadComplete,
-        onCancel: createAnn.onUploadCancel
+        uploadSuccess: composeAnn.onUploadComplete,
+        onCancel: composeAnn.onUploadCancel
       }
-      fileUpload.createFineUploadInstance(options)
+      fileUpload.createFineUploadInstance(options,function(data){
+        /*angular.forEach(composeAnn.announcement.attachments, function (attachment, key) {
```
Comment (kochhar) [inline src/app/private/scripts/controllers/announcement/createAnnouncementController.js]: please delete commented code.

File: src/test/spec/app/private/scripts/controllers/announcement/createAnnouncementController.js [hunk 18]
```diff
@@ -41,7 +40,7 @@ describe('Controller: createAnnouncementCtrl', function () {
     deferred = _$q_.defer()
     spyOn(toasterService, 'success').and.callThrough()
     spyOn(toasterService, 'error').and.callThrough()
-    spyOn(announcementService, 'getDefinitions').and.returnValue(deferred.promise)
+    //  spyOn(announcementAdapter, 'getDefinitions').and.returnValue(deferred.promise)
```
Comment (kochhar) [inline src/test/spec/app/private/scripts/controllers/announcement/createAnnouncementController.js]: please delete commented code.

File: src/app/private/scripts/telemetry/TelemetryService.js [hunk 19]
```diff
@@ -1,309 +1,317 @@
 TelemetryService = {
```
Comment (kochhar) [inline src/app/private/scripts/telemetry/TelemetryService.js]: please revert all changes in this file and bring them in via a new PR.

File: src/app/private/scripts/services/adapters/announcementAdapter.js [hunk 20]
```diff
@@ -229,13 +236,46 @@ angular.module('playerApp')
...
+     /**
+       * @method verifyAnnouncementData
+       * @desc Verify announcement data is filled before heading to next step
+       * @memberOf Services.announcementService
+       * @param {string}  stepNumber - step number
+       * @param {object}  announcement - announcement
+       * @returns {boolean} returns boolean
+       * @instance
+       */
+      this.verifyAnnouncementData = function (stepNumber, announcement) {
```
Comment (kochhar) [inline src/app/private/scripts/services/adapters/announcementAdapter.js]: the announcement adapter cannot be tied to the steps of the creation flow. please remove all references to step number.

File: src/app/private/scripts/controllers/announcement/resendAnnouncementController.js [hunk 21]
```diff
@@ -0,0 +1,420 @@
+'use strict'
```
Comment (kochhar) [inline src/app/private/scripts/controllers/announcement/resendAnnouncementController.js]: this controller is duplicating code in the sendAnnouncementController. please remove all duplicated code.

File: src/app/private/scripts/routes/announcementRoute.js [hunk 22]
```diff
@@ -89,4 +89,133 @@ angular.module('playerApp')
           $rootScope.homeActive = ''
         }
       })
-  })
+      .state('announcementCreate', {
```
Comment (kochhar) [inline src/app/private/scripts/routes/announcementRoute.js]: please have this file reviewed by @harishgilimi

**Comment by kochhar** [comment]
@niteshkesarkar please fix code duplication issues. please fix duplicate resendController.

**Pull request merged**